### PR TITLE
Nft transfer

### PR DESCRIPTION
**Pull request by mindcarver**
1. basically complete the nft-transfer contract
2. need to test erc1155 contract
3. need to test transfer function
4. need to test recvpacket function
5. need to test ack packet function
6. continue to improve the contract based on the test questions

### CODE REVIEW

File: contracts/libraries/30-nft-transfer/Strings.sol [hunk 1]
```diff
@@ -0,0 +1,683 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
```
Comment (chengwenxi) [inline contracts/libraries/30-nft-transfer/Strings.sol:2]: Can we find it on npm?

File: contracts/core/apps/30-nft-transfer/ERC1155Bank.sol [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    function transferFrom(address from, address to, uint256 id, uint256 amount, bytes memory data) public virtual override{
+        super.safeTransferFrom(from, to, id, amount, data);
+    }
+
+    function balanceOfAddr(address account, uint256 id) public view virtual returns (uint256) {
+        super.balanceOf(account, id);
+    }
+
+
```
Comment (chengwenxi) [inline contracts/core/apps/30-nft-transfer/ERC1155Bank.sol:31]: delete

File: contracts/core/apps/uu/Bytes.sol [hunk 3]
```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (chengwenxi) [inline contracts/core/apps/uu/Bytes.sol:1]: move to `libraries/utils`

File: contracts/core/apps/uu/Strings.sol [hunk 4]
```diff
@@ -0,0 +1,681 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (chengwenxi) [inline contracts/core/apps/uu/Strings.sol:1]: ditto.

File: contracts/core/apps/30-nft-transfer/ERC1155Bank.sol [hunk 5]
```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.6.8;
+pragma experimental ABIEncoderV2;
+
+import "openzeppelin-solidity/contracts/token/ERC1155/ERC1155.sol";
+import "../../../interfaces/IERC1155Bank.sol";
+
+import "hardhat/console.sol";
```
Comment (dreamer-zq) [inline contracts/core/apps/30-nft-transfer/ERC1155Bank.sol:8]: remove test code

File: contracts/proto/NftTransfer.sol [hunk 6]
```diff
@@ -0,0 +1,487 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (chengwenxi) [inline contracts/proto/NftTransfer.sol:1]: move to `libraries/apps/nft`

File: contracts/core/apps/30-nft-transfer/Transfer.sol [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+            sender: Bytes.addressToString(msg.sender),
+            receiver: transferData.receiver,
+            awayFromOrigin :  awayFromOrigin
+            })
+        );
+
+        // send packet
+        PacketTypes.Packet memory pac = PacketTypes.Packet({
+            sequence : packet.getNextSequenceSend(sourceChain, transferData.destChain),
+            port : "NFT",
```
Comment (chengwenxi) [inline contracts/core/apps/30-nft-transfer/Transfer.sol:88]: define port as constant

**Review by chengwenxi — COMMENTED** [review]
pls format each contract and annotate each function.

File: contracts/core/apps/30-nft-transfer/Transfer.sol [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+    */
+    mapping(uint256 => string) public classMapping;
+    mapping(uint256 => string) public idMapping;
+    mapping(uint256 => string) public uriMapping;
+
+    struct NftMapValue {
+        string  class;
+        string  id;
+        string  uri;
+    }
```
[the remark this answers is not in the document]
Comment (mindcarver) [reply contracts/core/apps/30-nft-transfer/Transfer.sol:51]: This is to use structure to reduce local variables and solve the problem of stack too deep

**Review by chengwenxi — APPROVED** [review]

**Review by dreamer-zq — APPROVED** [review]

**Pull request merged**